extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/WriteMongoP.java

```diff
@@ -0,0 +1,589 @@
+
+            ArrayList<I> items = new ArrayList<>();
+            for (Object item : inbox) {
+                items.add((I) item);
+                if (items.size() >= MAX_BATCH_SIZE) {
+                    break;
+                }
+            }
+            @SuppressWarnings("DataFlowIssue")
+            Map<MongoCollectionKey, List<I>> itemsPerCollection = items
+                    .stream()
+                    .map(e -> tuple2(collectionPicker.pick(e), e))
+                    .collect(groupingBy(Tuple2::f0, mapping(Tuple2::getValue, toList())));
+
+            for (MongoCollectionKey collectionKey : itemsPerCollection.keySet()) {
+                MongoCollection<I> collection = collectionKey.get(connection.client(), documentType);
+
+                Set<Object> docsFound = queryForExistingIds(items, collection);
+                List<WriteModel<I>> writes = new ArrayList<>();
+
+                for (I item : itemsPerCollection.get(collectionKey)) {
+                    Object id = documentIdentityFn.apply(item);
+
+                    WriteModel<I> write;
+                    if (docsFound.contains(id)) {
+                        Bson filter = eq(documentIdentityFieldName, id);
+                        write = new ReplaceOneModel<>(filter, item, replaceOptions);
+                    } else {
+                        write = new InsertOneModel<>(item);
+                    }
+                    writes.add(write);
+                }
+                if (transactionUtility.usesTransactionLifecycle()) {
+                    checkNotNull(mongoTransaction, "there is no active transaction");
+                    mongoTransaction.addWrites(collectionKey, writes);
+                } else {
+                    collection.bulkWrite(writes);
+                }
+            }
+
+            for (int i = 0; i < items.size(); i++) {
+                inbox.remove();
+            }
+        } catch (Exception e) {
+            logger.info("Unable to process Mongo Sink: " + e.getMessage());
+            // not removing from inbox, so it will be retried
+        }
+    }
+
+    private Set<Object> queryForExistingIds(ArrayList<I> items, MongoCollection<?> collection) {
+        List<Object> identityValues = items.stream()
+                                           .map(documentIdentityFn)
+                                           .filter(Objects::nonNull)
+                                           .collect(toList());
+        Iterable<Document> foundIds = collection.aggregate(
+                asList(project(include("_id")), match(in("_id", identityValues))),
+                Document.class);
+        Set<Object> docsFound = new HashSet<>();
+        for (Document document : foundIds) {
+            docsFound.add(document.get("_id"));
+        }
+        return docsFound;
+    }
+
+    @Override
+    public void close() {
+        connection.close();
+        transactionUtility.close();
```

Comment:
if the first close fails, the 2nd is not called. Use e.g. a try with resources block or something.